feat(rpc)!: Update `Emitter::mempool` to support `evicted_at`

LagginTimes opened the pull request:
Fixes #1740

### Description

Work for this began as part of #1811, based on this [comment](https://github.com/bitcoindevkit/bdk/pull/1811#discussion_r1945941691).

`Emitter::mempool` now returns a `MempoolEvent` which provides the data for tracking `evicted_at`:

```
pub struct MempoolEvent {
    /// Unemitted transactions or transactions with ancestors that are unseen by the receiver.
    ///
    /// To understand the second condition, consider a receiver which filters transactions based on
    /// whether it alters the UTXO set of tracked script pubkeys. If an emitted mempool transaction
    /// spends a tracked UTXO which is confirmed at height `h`, but the receiver has only seen up to
    /// block of height `h-1`, we want to re-emit this transaction until the receiver has seen the
    /// block at height `h`.
    pub new_txs: Vec<(Transaction, u64)>,

    /// [`Txid`]s of all transactions that have been evicted from mempool.
    pub evicted_txids: HashSet<Txid>,

    /// The latest timestamp of when a transaction entered the mempool.
    ///
    /// This is useful for setting the timestamp for evicted transactions.
    pub latest_update_time: u64,
}
```

### Changelog notice

* Change `Emitter::mempool` to return `MempoolEvent`s which contain mempool-eviction data.
* Change `Emitter::client` to have more relaxed generic bounds. `C: Deref, C::Target: RpcApi` are the new bounds.
* Add conversion impls for `CanonicalTx` to `Txid`/`Arc<Transaction>`.
* Add `ChainPosition::is_unconfirmed` method.

### Checklists

#### All Submissions:

* [x] I've signed all my commits
* [x] I followed the [contribution guidelines](https://github.com/bitcoindevkit/bdk/blob/master/CONTRIBUTING.md)
* [x] I ran `cargo fmt` and `cargo clippy` before committing

#### New Features:

* [x] I've added tests for the new feature
* [x] I've added docs for the new feature

#### Bugfixes:

* [x] This pull request breaks the existing API
* [x] I've added tests to reproduce the issue which are now passing
* [x] I'm linking the issue being fixed by this PR

ValuedMammal committed:
refactor(rpc)!: update `mempool` interface and test code

evanlinjin commented:
Instead of rebasing, it's probably best to `git switch -c <new-branch>` from #1839 and cherry-pick on the new commits.

evanlinjin commented on crates/bitcoind_rpc/src/lib.rs:
```diff
@@ -84,24 +96,54 @@ impl<'c, C: bitcoincore_rpc::RpcApi> Emitter<'c, C> {
+        let raw_mempool = client.get_raw_mempool_verbose()?;
+        let raw_mempool_txids: HashSet<Txid> = raw_mempool.keys().copied().collect();
+
+        // Determine the latest block height from the raw mempool data.
```
Why would you determine the latest block height from the mempool data?

evanlinjin commented on crates/bitcoind_rpc/src/lib.rs:
```diff
@@ -84,24 +96,41 @@ impl<'c, C: bitcoincore_rpc::RpcApi> Emitter<'c, C> {
+        let raw_mempool_txids: HashSet<Txid> = raw_mempool.keys().copied().collect();
+
+        // Determine if height matches last emitted block.
+        let at_tip = client.get_block_count()? == self.last_cp.height() as u64;
```
What if it's the same height but different block due to reorg?

evanlinjin commented on crates/bitcoind_rpc/src/lib.rs:
```diff
@@ -112,23 +141,40 @@ impl<'c, C: bitcoincore_rpc::RpcApi> Emitter<'c, C> {
+                new_txs
+                    .iter()
+                    .map(|(tx, _)| tx.compute_txid())
+                    .collect::<Vec<_>>(),
```
I don't think a collect is necessary here.

evanlinjin commented on crates/bitcoind_rpc/tests/test_emitter.rs:
```diff
@@ -731,3 +744,107 @@ fn no_agreement_point() -> anyhow::Result<()> {
+    // Update graph with evicted tx.
+    for txid in mempool_event.evicted_txids {
+        let _ = graph.insert_evicted_at(txid, seen_at);
+    }
```
We need to filter out irrelevant `evicted_at`s. I.e. if the txid is not part of our wallet.

evanlinjin reviewed (changes requested):
Documentation is outdated, but I understand this is WIP.

evanlinjin commented on crates/bitcoind_rpc/src/lib.rs:
```diff
@@ -43,6 +44,9 @@ pub struct Emitter<'c, C> {
     /// there has been a reorg since our last mempool emission.
     last_mempool_tip: Option<u32>,
+
+    /// Expected mempool txs. TODO: Docs.
```
Are you planning to address this TODO?

LagginTimes replied:
Oops, looks like I goofed the rebase and undid all my changes to `lib.rs`... should be fixed now.

evanlinjin commented on crates/bitcoind_rpc/src/lib.rs:
```diff
@@ -53,18 +58,34 @@ impl<'c, C: bitcoincore_rpc::RpcApi> Emitter<'c, C> {
-    pub fn new(client: &'c C, last_cp: CheckPoint, start_height: u32) -> Self {
+    ///
+    /// `expected_mempool_txids` is the initial set of unconfirmed txids. Once at tip, any that are
+    /// no longer in mempool are marked evicted.
```
```suggestion
    /// `expected_mempool_txids` is the initial set of unconfirmed txids provided by the wallet. This
    /// allows the [`Emitter`] to inform the wallet about relevant mempool evictions.
```

evanlinjin commented on crates/bitcoind_rpc/src/lib.rs:
```diff
@@ -43,6 +44,10 @@ pub struct Emitter<'c, C> {
     last_mempool_tip: Option<u32>,
+
+    /// Unconfirmed txids that are expected to appear in mempool. This is used to determine if any
+    /// known txids have been evicted.
```
```suggestion
/// A set of txids that we currently believe are still in the mempool.
///
/// This is used to detect mempool evictions by comparing the set against the latest
/// mempool snapshot from bitcoind. Any txid in this set that is missing from the
/// snapshot is considered evicted.
///
/// When the emitter emits a block, confirmed txids are removed from this set. This prevents confirmed transactions from being
/// mistakenly marked with an `evicted_at` timestamp.
```

evanlinjin commented on crates/bitcoind_rpc/src/lib.rs:
```diff
@@ -53,18 +58,34 @@ impl<'c, C: bitcoincore_rpc::RpcApi> Emitter<'c, C> {
 
-    /// Emit mempool transactions, alongside their first-seen unix timestamps.
+    /// Emit mempool transactions and any evicted [`Txid`]s. Returns a `latest_update_time` which is
+    /// used for setting the timestamp for evicted transactions.
```
```suggestion
    /// Emit mempool transactions and any evicted [`Txid`]s.
```

I don't think this belongs here.

evanlinjin commented on crates/bitcoind_rpc/src/lib.rs:
```diff
@@ -112,14 +158,17 @@ impl<'c, C: bitcoincore_rpc::RpcApi> Emitter<'c, C> {
+                            if at_tip {
+                                evicted_txids.insert(txid);
+                            }
+                            return None;
```
It's also possible that a new block is found while we were looping through the mempool txs.

If this tx is confirmed in the new block (and therefore no longer exists in the mempool), the correct behavior is just to ignore it otherwise we may emit a false `evicted_at` for that tx here.

If the tx is actually evicted from the mempool after the `client.get_raw_mempool_verbose()` call, we will still detect the eviction will another `.mempool` call. 

I think we should retain the old logic.

evanlinjin reviewed (changes requested):
I haven't finished reviewing yet. Some thoughts:

1. We probably need something like `insert_relevant_evicted_ats(&mut self, evicted_txs: impl IntoIterator<Item = (Txid, u64)>)` on `TxGraph` for convenience.

2. The old best-effort-avoid-reemitting-mempool-txs logic might actually be replaceable by caching `Arc<Transaction>` in `expected_mempool_txids`.

    ```rust
    prior_mempool_txs: HashMap<Txid, Arc<Transaction>>,
    ```

    The idea is we always emit the whole mempool but it is still memory-efficient.

For `2.`, I guess we can tackle that in a separate PR.

evanlinjin commented on crates/bitcoind_rpc/tests/test_emitter.rs:
```diff
@@ -731,3 +744,109 @@ fn no_agreement_point() -> anyhow::Result<()> {
 }
+
+#[test]
+fn test_expect_tx_evicted() -> anyhow::Result<()> {
```
Would you be able to add a brief summary at the top of the test (e.g. as a doc comment) explaining what the test is validating?

Without a summary, it’s a bit hard to grasp the intent of the test without reading through every line and trying to hold the context in mind. A short explanation would really help with readability, especially for larger tests like this one.

evanlinjin reviewed (commented):
The logic looks solid.

There’s room for improvement in documentation clarity across tests, inline code comments, and public doc comments:

- **Code comments**: Focus on the _why_ behind the logic, rather than describing _what_ the code is doing—since that’s usually self-evident from the code itself.
- **Tests**: For larger tests, consider adding a summary doc comment at the top. This helps readers quickly understand what the test is verifying without having to trace through each line and reconstruct the intent.
- **Public doc comments**: Prioritize what’s important for the _caller_ to know—avoid going into implementation details unless they directly impact usage.

evanlinjin commented on crates/bitcoind_rpc/src/lib.rs:
```diff
@@ -84,24 +105,49 @@ impl<'c, C: bitcoincore_rpc::RpcApi> Emitter<'c, C> {
                         *latest_time = tx_time;
                     }
-
                     // Avoid emitting transactions that are already emitted if we can guarantee
```
Could you help me replace this code comment with this?

```rust
                    // Best-effort check to avoid re-emitting transactions we've already emitted.
                    //
                    // Complete suppression isn't possible, since a transaction may spend outputs
                    // owned by the wallet. To determine if such a transaction is relevant, we must
                    // have already seen its ancestor(s) that contain the spent prevouts.
                    //
                    // Fortunately, bitcoind provides the block height at which the transaction
                    // entered the mempool. If we've already emitted that block height, we can
                    // reasonably assume the receiver has seen all ancestor transactions.
```

LagginTimes committed:
feat(graph): add convenience function for inserting relevant `evicted_at`s

evanlinjin commented on crates/chain/src/tx_graph.rs:
```diff
@@ -777,6 +777,27 @@ impl<A: Anchor> TxGraph<A> {
+        let mut changeset = ChangeSet::default();
+        for txid in txids {
+            // Only record evictions for transactions the graph is tracking.
+            if self.get_tx_node(txid).is_some() {
```
```suggestion
            if self.txs.contains(txid) {
```

evanlinjin commented on examples/example_bitcoind_rpc_polling/src/main.rs:
```diff
@@ -287,7 +286,13 @@ fn main() -> anyhow::Result<()> {
+                            graph_changeset.merge(
+                                graph.insert_evicted_at(txid, mempool_txs.latest_update_time),
+                            );
+                        }
```
Needs update to use `insert_relevant_evicted_ats`

evanlinjin commented on crates/chain/src/indexed_tx_graph.rs:
```diff
@@ -145,6 +145,23 @@ where
+    /// The `evicted_at` timestamp represents the last known time when the transaction was observed
+    /// to be missing from the mempool. If `txid` was previously recorded with an earlier
+    /// `evicted_at` value, it is updated only if the new value is greater.
+    pub fn insert_relevant_evicted_ats(
```
For consistency with other methods in `IndexedTxGraph`, I propose the following:

1. Rename the method to `batch_insert_relevant_evicted_at`.
2. Update the method signature to:  
   `batch_insert_relevant_evicted_at(&mut self, evicted_ats: impl IntoIterator<Item = (Txid, u64)>)`

This API more directly reflects the internal logic — applying individual `evicted_at` timestamps to each `txid`.

To make this integrate better with `MempoolEvent`, we could either:

1. Change the `MempoolEvent::evicted_txids` field to a `HashMap<Txid, u64>`, and remove the `latest_update_time` field.
2. Add a helper method to `MempoolEvent` that returns an `impl ExactSizeIterator<Item = (Txid, u64)>`.

Option 2 might be more direct and would require fewer changes.

Let me know what you think.

LagginTimes replied:
I think option 2 looks good to me. Let me know if the implementation looks good to you.

evanlinjin reviewed (commented):
Looks good, just a final request.

evanlinjin commented on crates/bitcoind_rpc/src/lib.rs:
```diff
@@ -84,15 +110,40 @@ impl<'c, C: bitcoincore_rpc::RpcApi> Emitter<'c, C> {
+        } else {
+            // Verify if block hash matches in case of re-org.
+            client.get_block_hash(height)? == self.last_cp.hash()
+        };
```
We want to make sure that the fetched mempool contents and the fetched tip are always guaranteed to be consistent with one another. I.e. in other words, txs returned by `get_raw_mempool_verbose` should not exist in the block of `get_block_hash`.

The current logic does not enforce this. There is a slim chance the node might find a new block between the `get_raw_mempool_verbose` call and the `get_block_hash` call. If this happens, the current logic will advertise confirmed transactions as evicted.

evanlinjin commented on crates/bitcoind_rpc/src/lib.rs:
```diff
@@ -53,18 +64,33 @@ impl<'c, C: bitcoincore_rpc::RpcApi> Emitter<'c, C> {
+    /// first-seen unix timestamps) that were emitted, and [`MempoolEvent::evicted_txids`] which are
+    /// any [`Txid`]s which were previously expected and are now missing from the mempool. Note that
+    /// [`Txid`]s are only evicted if the emitter is at the chain tip with the same height and hash
+    /// as the best block from the RPC.
```
"Previously expected" can be a bit unclear. Suggestion:  
**"Were previously seen in the mempool and are now missing."**

Also, the content in the `Note that...` sentence is important for the caller to understand, but it's not stated clearly enough.  
Specifically, the caller should be made aware that if they don't call `next_block`, they will never be informed about transaction evictions.

evanlinjin reviewed (commented):
Apologies for the incremental reviews — it's been eye-opening for me as well how tricky the `Emitter` is to reason about. Let's aim to simplify it in follow-up PRs.

One thing I’ve come to realize: the `next_header` method is inconsistent with the eviction-detection logic we're introducing via `next_block` and `mempool`. I suggest removing `next_header` entirely — everything it does can already be achieved through `next_block`.

evanlinjin commented on examples/example_bitcoind_rpc_polling/src/main.rs:
```diff
@@ -141,7 +139,7 @@ fn main() -> anyhow::Result<()> {
             let chain_tip = chain.lock().unwrap().tip();
             let rpc_client = rpc_args.new_client()?;
-            let mut emitter = Emitter::new(&rpc_client, chain_tip, fallback_height);
+            let mut emitter = Emitter::new(&rpc_client, chain_tip, fallback_height, HashSet::new());
```
We should probably provide the initial unconfirmed contents here.

evanlinjin replied:
I fixed this. Also made some API changes to reduce the friction of populating `expected_mempool_txids` from `bdk_chain`.

evanlinjin committed:
feat(chain): Add convenience conversions for `CanonicalTx`

* `From<CanonicalTx> for Txid`
* `From<CanonicalTx> for Arc<Transaction>`

Also added a convenience method `ChainPosition::is_unconfirmed`.

These are intended to simplify the calls needed to populate the
`expected_mempool_txids` field of `Emitter::new`.

evanlinjin committed:
feat(bitcoind_rpc)!: Reduce friction of `Emitter` API.

* Change signature of `Emitter::new` so that `expected_mempool_txids`
  can be more easily constructed from `TxGraph` methods.
* Change generic bounds of `C` within `Emitter<C>` to be `C: DeRef,
  C::Target: RpcApi`. This allows the caller to have `Arc<Client>` as
  `C` and does not force to caller to hold a lifetimed reference.

evanlinjin reviewed (approved):
self-ACK e1c53d737b396474cc1015183f3bb1a43afd12d0

ValuedMammal commented:
@LagginTimes Can you rebase on master and also squash 6e1178ca into the previous commit?

ValuedMammal reviewed (approved):
ACK 8513d83a2ef2fe748e9007984187a7f08d256ee6

ValuedMammal commented:
Thanks @LagginTimes 

I didn't mean to become author on 28ef7c9a 😅
